runtime/doc/lua.txt

```diff
@@ -1772,6 +1756,29 @@ vim.schedule_wrap({fn})                                  *vim.schedule_wrap()*
       • |vim.schedule()|
       • |vim.in_fast_event()|
 
+vim.str_byteindex({s}, {index}, {opts})                  *vim.str_byteindex()*
+    Convert UTF-32 or UTF-16 {index} to byte index. {opts} can be a table
+    containing two properties:i
+    • an "encoding" property of "utf-8", "utf-16" or "utf-32"
+    • an "error" property of true or false, if error is false then then an out
+      of range index will return byte length instead of throwing an error.
+
+    {opts} can also be supplied as a boolean, true for utf-16 and false for
+    utf-32. If {opts} is not supplied, it defaults to false (UTF-32). Returns
+    the byte index.
+
+    Invalid UTF-8 and NUL is treated like in |vim.str_utfindex()|. An {index}
+    in the middle of a UTF-16 sequence is rounded upwards to the end of that
+    sequence.
+
+    Parameters: ~
+      • {s}      (`string`)
+      • {index}  (`integer`)
+      • {opts}   (`boolean?|{ encoding: "utf-8"|"utf-16"|"utf-32", error?: boolean }?`)
```

Comment:
Is there a reason of having a specific `error` opt instead of using `pcall`?

Reply:
the idea was to avoid pcalls and handle the errors gracefully to replace ```vim.lsp.util._str_byteindex_enc()```

Reply:
To me handling the error gracefully means detecting what the error is and acting appropriately. From what I can tell this change just ignores the error completely. I'm not sure how that is more graceful than using `pcall`?

Reply:
@clason wdyt?

Reply:
> The inspiration for the interface was from the recent `get_parser()` changes

The `error` flag added to get_parser is a temporary workaround. See `:help dev-error-patterns` for recommended ways to pass errors/failure.

Reply:
> > The inspiration for the interface was from the recent `get_parser()` changes
> 
> The `error` flag added to `get_parser` is a temporary workaround, probably not relevant here. Instead see `:help dev-error-patterns` for recommended ways to pass errors/failure.

I have read that documentation. The point of the flag is essentially choosing between behaviour 3 or a fallback value.

I could go with option 1, and provide
```i, nil``` in a success case and ```#s, "index out of range"``` in a failure case. But given that the function throws at the moment, there may be other callers depending on the throw behaviour

Reply:
> The point of the flag is essentially choosing between behaviour 3

The flag isn't a pattern that we want to continue. So unless it's a temporary bridge that we can't avoid, I'd suggest just throwing an error if that's how it was already done and we think changing that would break consumers.

Reply:
> > The point of the flag is essentially choosing between behaviour 3
> 
> The flag isn't a pattern that we want to continue. So unless it's a temporary bridge that we can't avoid, I'd suggest just throwing an error if that's how it was already done and we think changing that would break consumers.

we need to be able to suppress the throw and provide a fallback to be able to replace the ```vim.lsp.util._str_byteindex_enc()```

What do you think about a fallback flag? I am trying to think of where we have a similar pattern for precedence

Reply:
The closest I can think of is an ```on_error``` callback function that would be similar to ```on_init``` or ```on_attach```

Reply:
What is the problem with using `pcall`?

Reply:
> What is the problem with using `pcall`?

The motivation for rewriting this was part of #25272 to ultimately be able to retire ```_str_byteindex_enc```
We don't want to have to call every LSP call to ```str_byteindex``` as that leads to overhead in callers code and  can mask errors.

``'
                                                          *lua-error-handling*
Lua functions may throw |lua-errors| for exceptional (unexpected) failures,
which you can handle with |pcall()|.
```

I would argue that an index being out of range isn't exceptional enough to throw

Reply:
> and can mask errors.

What errors? The only error this can really throw is argument type or out of index.

Out of index seems exceptional to me. Using pcall seems like the most ideomatic route. This function just isn't complicated enough to justify doing anything else.

Reply:
> > The motivation for rewriting this was part of #25272
> 
> I can't find anything in that issue that specifically mentions anything about avoiding pcall. All it says is "Consider updating those functions to accept an encoding instead."
> 

But there is a bounds check/fallback in the existing ```_str_byteindex_enc``` check here
https://github.com/neovim/neovim/blob/8ef3dd3afa20024f439d8b7a69f316069c1909d3/runtime/lua/vim/lsp/util.lua#L174-L178

Additionally, the current bounds check isn't sufficient if you take the below code
```lua
str = '世e'

vim.print(vim.lsp.util._str_byteindex_enc(str, 1, 'utf-16'))
vim.print(vim.lsp.util._str_byteindex_enc(str, 2, 'utf-16'))
-- vim.print(vim.lsp.util._str_byteindex_enc(str, 3, 'utf-16'))
-- vim.print(vim.lsp.util._str_byteindex_enc(str, 4, 'utf-16'))
vim.print(vim.lsp.util._str_byteindex_enc(str, 5, 'utf-16'))
```

The commented lines will currently throw, the others will return 2, 4, and 4 (and any index above 4 would return 4 too)

so we would need to be able to use that bounds check in a higher level implementation too or we would just need another wrapper function

Reply:
The functions is called like 4 times. Why not just pull the bounds checking outside? Is it even completely necessary in each cace?

Reply:
I like the idea of stopping it throwing and moving it to a ```retval, errmsg``` return

Then the lsp code can just use 
```lua
vim.str_byteindex(string, 10, { encoding = 'utf-16' }) or #string
```

But is that considered breaking for a function that used to throw previously

Reply:
Yes. Please just use `pcall`, or just throw the error i the LSP code.